backends/simplego/dotgeneral.go

```diff
@@ -177,29 +177,69 @@ func execNormalizedDotGeneralGeneric[T PODNumericConstraints](lhs, rhs, output *
 	lhsFlat := lhs.flat.([]T)
 	rhsFlat := rhs.flat.([]T)
 	outputFlat := output.flat.([]T)
-	var lhsIdx, rhsIdx, outputIdx int
-	batchSize := lhs.shape.Dimensions[0] // same as rhs'.
+
+	batchSize := lhs.shape.Dimensions[0]
 	lhsCrossSize := lhs.shape.Dimensions[1]
 	rhsCrossSize := rhs.shape.Dimensions[1]
-	contractingSize := lhs.shape.Dimensions[2] // same as rhs'.
-	rhsBatchStride := contractingSize * rhsCrossSize
-	for range batchSize {
-		for range lhsCrossSize {
-			rhsBatchStartIdx := rhsIdx
-			for range rhsCrossSize {
-				lhsRowStartIdx := lhsIdx
-				for range contractingSize {
-					outputFlat[outputIdx] += lhsFlat[lhsIdx] * rhsFlat[rhsIdx]
-					lhsIdx++
-					rhsIdx++
+	contractingSize := lhs.shape.Dimensions[2]
+
+	// Pre-compute strides to avoid repeated calculations
+	lhsBatchStride := lhsCrossSize * contractingSize
+	rhsBatchStride := rhsCrossSize * contractingSize
+	outputBatchStride := lhsCrossSize * rhsCrossSize
+
+	// Zero out the output buffer in one go for better memory efficiency
+	for i := range outputFlat {
+		outputFlat[i] = 0
+	}
+
+	// Cache block sizes - adjust based on typical matrix sizes and CPU cache
+	const blockSize = 64 // Tune this based on your typical workload and L1 cache size
+
+	for batchIdx := 0; batchIdx < batchSize; batchIdx++ {
+		lhsBaseIdx := batchIdx * lhsBatchStride
+		rhsBaseIdx := batchIdx * rhsBatchStride
+		outputBaseIdx := batchIdx * outputBatchStride
+
+		// Use blocking to improve cache locality
+		for ii := 0; ii < lhsCrossSize; ii += blockSize {
+			iEnd := min(ii+blockSize, lhsCrossSize)
+
+			for jj := 0; jj < rhsCrossSize; jj += blockSize {
+				jEnd := min(jj+blockSize, rhsCrossSize)
+
+				for kk := 0; kk < contractingSize; kk += blockSize {
+					kEnd := min(kk+blockSize, contractingSize)
+
+					// Process the current block
+					for i := ii; i < iEnd; i++ {
```

Comment:
Maybe rename "i" -- having "i" and "ii" is a bit confusing.

Because there are so many iterators, I'd start giving a long name to them: idxLhsCross(ii), idxRhsCross (jj), idxContracting (kk), idxBlock (i)